agera/src/main/java/com/google/android/agera/Observables.java

```diff
@@ -138,31 +139,77 @@ public static UpdateDispatcher updateDispatcher(
     return new AsyncUpdateDispatcher(activationHandler);
   }
 
-  private static final class CompositeObservable extends BaseObservable implements Updatable {
+  private static final class CompositeObservable implements Observable {
+    @NonNull
+    private static final Updatable[] NO_UPDATABLES = new Updatable[0];
     @NonNull
     private final Observable[] observables;
+    @NonNull
+    private Updatable[] updatables;
 
     CompositeObservable(@NonNull final Observable... observables) {
       this.observables = observables;
+      this.updatables = NO_UPDATABLES;
     }
 
     @Override
-    protected void observableActivated() {
-      for (final Observable observable : observables) {
-        observable.addUpdatable(this);
+    public synchronized void addUpdatable(@NonNull final Updatable updatable) {
+      boolean added = false;
+      Updatable localUpdatable = null;
+      for (int index = 0; index < updatables.length; index += 2) {
+        if (updatables[index] == updatable) {
+          throw new IllegalStateException("Updatable already added, cannot add.");
+        }
+        if (updatables[index] == null && !added) {
+          updatables[index] = updatable;
+          updatables[index + 1] = new WrapperUpdatable(updatable);
+          localUpdatable = updatables[index + 1];
+          added = true;
+        }
+      }
+      if (!added) {
+        final int newIndex = updatables.length;
+        updatables = Arrays.copyOf(updatables,
+            Math.max(newIndex * 2, newIndex + 2));
```

Comment:
this line confusing, could you replace it with `newIndex < 2 ? 2 : newIndex * 2`